dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+        set ``vectorize=True``).
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
+    output_dtypes : dtype or list of dtypes
+        dtype or list of output dtypes.
+    output_sizes : dict, optional
+        Optional mapping from dimension names to sizes for outputs. Only used if
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+
+    Returns
+    -------
+    Single dask.array.Array or tuple of dask.array.Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    ... import numpy as np
+    ... def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    ... a = da.random.normal(size=(10,20,30), chunks=5)
+    ... mean, std = da.apply_gufunc(stats, "(i)->(),()", a, output_dtypes=2*(a.dtype,))
+    ... mean.compute()  # doctest: +SKIP
+
+    >>> import dask.array as da
+    ... import numpy as np
+    ... def outer_product(x, y):
+    ...     return np.einsum("...i,...j->...ij", x, y)
+    ... a = da.random.normal(size=(   20,30), chunks=5)
+    ... b = da.random.normal(size=(10, 1,40), chunks=10)
+    ... c = da.apply_gufunc(outer_product, "(i),(j)->(i,j)", a, b, output_dtypes=a.dtype)
+    ... c.compute()  # doctest: +SKIP
+
+    References
+    ----------
+    .. [1] http://docs.scipy.org/doc/numpy/reference/ufuncs.html
+    .. [2] http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+    """
+    output_dtypes = kwargs.pop("output_dtypes", None)
+    output_sizes = kwargs.pop("output_sizes", None)
+    vectorize = kwargs.pop("vectorize", None)
+    concatenate = kwargs.pop("concatenate", True)
+    if output_dtypes is None:
+        raise ValueError("Must specify `output_dtypes` of output array(s)")
+
+    # Input processing:
+    ## Signature
+    if isinstance(signature, str):
+        core_input_dimss, core_output_dimss = parse_signature(signature)
+    else:
+        core_input_dimss, core_output_dimss = signature
+
+    ## Determine nout
+    nout = None if not isinstance(core_output_dimss, list) else len(core_output_dimss)
+
+    ## Assert output_dtypes
+    if nout is not None:
+        if not (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
+            raise ValueError("Must specify tuple of dtypes for `output_dtypes` for function with multiple outputs")
+        if len(output_dtypes) != nout:
+            raise ValueError("Must specify tuple of %d dtypes for `output_dtypes` for function with multiple outputs" % nout)
+    if nout is None and (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
+        raise ValueError("Must specify single dtype for `output_dtypes` for function with one output")
```

Comment:
Just checking, but do we test for this behavior?  (I haven't gotten down to that yet)